modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java

```diff
@@ -396,90 +390,83 @@ public CompletableFuture<Ignite> start(@Language("HOCON") @Nullable String cfg)
+                        CompletableFuture<Void> recoveryFuture = RecoveryCompletionFutureFactory.create(
+                                clusterCfgMgr,
+                                fut -> new ConfigurationCatchUpListener(cfgStorage, fut, LOG)
+                        );
+
+                        return CompletableFuture.allOf(notifyConfigurationListeners(), recoveryFuture);
+                    })
+                    // Signal that local recovery is complete and the node is ready to join the cluster.
+                    .thenCompose(v -> cmgMgr.onJoinReady())
+                    .thenAccept(v -> {
+                        try {
+                            // Transfer the node to the STARTED state.
+                            lifecycleManager.onStartComplete();
+                        } catch (NodeStoppingException e) {
+                            throw new CompletionException(e);
+                        }
+                    })
+                    .handle((v, e) -> {
+                        if (e != null) {
+                            throw handleStartException(e);
+                        }
+
+                        return this;
+                    });
         } catch (Exception e) {
-            String errMsg = "Unable to start node=[" + name + "].";
+            throw handleStartException(e);
+        }
+    }
 
-            LOG.error(errMsg, e);
+    private RuntimeException handleStartException(Throwable e) {
+        String errMsg = "Unable to start node=[" + name + "].";
 
-            doStopNode(startedComponents);
+        LOG.error(errMsg, e);
 
-            throw new IgniteException(errMsg, e);
-        }
-    }
+        lifecycleManager.stopAllComponents();
 
-    /**
-     * Awaits for a permission to join the cluster, i.e. node join response from Cluster Management group.
-     * After the completion of this method, the node is considered as validated.
-     */
-    private void waitForJoinPermission() {
-        // TODO: implement, see https://issues.apache.org/jira/browse/IGNITE-16472
+        return new IgniteException(errMsg, e);
     }
 
     /**
      * Stops ignite node.
      */
     public void stop() {
-        if (status.getAndSet(Status.STOPPING) == Status.STARTED) {
-            doStopNode(List.of(
-                    longJvmPauseDetector,
-                    vaultMgr,
-                    nodeCfgMgr,
-                    clusterSvc,
-                    computeComponent,
-                    raftMgr,
-                    txManager,
-                    metaStorageMgr,
-                    clusterCfgMgr,
-                    cmgMgr,
-                    baselineMgr,
-                    dataStorageMgr,
-                    distributedTblMgr,
-                    qryEngine,
-                    restComponent,
-                    clientHandlerModule,
-                    nettyBootstrapFactory
-            ));
-        }
+        lifecycleManager.stopNode();
```

Comment:
We should only stop node if the status was STARTED. 
So let's use 
```
        if (status.getAndSet(Status.STOPPING) == Status.STARTED) {
            stopAllComponents();
        }
```
as stopNode impementation.

Reply:
I've fixed this comment, but then I remembered that I actually did this on purpose: it is possible that a node may hang during join in which case it will never acquire the STARTED state. This will make it impossible to stop it in "afterEach" in tests, for example